src/dtls.c

```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+                             pskInfo->pv.minor != pv.minor))
+        return VERSION_ERROR;
+    if (IsAtLeastTLSv1_3(pv))
+        *isTls13 = 1;
+    else
+        *isTls13 = 0;
+
+    return 0;
+}
+
+static int CopyExtensions(TLSX* src, TLSX** dst, void* heap)
+{
+    /* Copy the following extensions
+     * * SupportedCurves */
+    TLSX* extension;
+    int ret;
+
+    extension = TLSX_Find(src, TLSX_SUPPORTED_GROUPS);
+    if (extension != NULL) {
+        SupportedCurve* curve;
+        curve = (SupportedCurve*)extension->data;
+        for (curve = (SupportedCurve*)extension->data; curve != NULL;
+                curve = curve->next) {
+            ret = TLSX_UseSupportedCurve(dst, curve->name, heap);
+            if (ret != WOLFSSL_SUCCESS)
+                return MEMORY_E;
+        }
+    }
+
+    return 0;
+}
+#endif
+
+static int SendStatelessReply(const WOLFSSL* ssl, WolfSSL_CH* ch, byte isTls13,
+                              PskInfo* pskInfo)
+{
+    int ret;
+    (void)isTls13;
+    (void)pskInfo;
+#ifdef WOLFSSL_DTLS13
+    if (isTls13) {
+#ifdef WOLFSSL_SEND_HRR_COOKIE
+        if (ch->cookie.size == 0) {
+            TLSX* parsedExts = NULL;
+            WolfSSL_ConstVector tlsx;
+            Suites suites;
+            word16 len;
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+            byte haveKS = 0;
+            byte usePSK = 0;
+            byte doKE = 0;
+#endif
+            CipherSuite cs;
+            CipherSpecs specs;
+
+            XMEMSET(&cs, 0, sizeof(cs));
+            XMEMSET(&specs, 0, sizeof(specs));
+
+            /* We need to echo the session ID sent by the client */
+            if (ch->sessionId.size > ID_LEN) {
+                /* Too large. We can't echo this. */
+                ERROR_OUT(INVALID_PARAMETER, dtls13_cleanup);
+            }
+
+            /* Hashes are reset in SendTls13ServerHello when sending a HRR */
+            ret = Dtls13HashHandshake((WOLFSSL*)ssl, ch->msg, ch->length);
+            if (ret != 0)
+                goto dtls13_cleanup;
+
+            /* Populate the suites struct to find a common ciphersuite */
+            XMEMSET(&suites, 0, sizeof(suites));
+            suites.suiteSz = ch->cipherSuite.size;
+            if ((suites.suiteSz % 2) != 0)
+                ERROR_OUT(INVALID_PARAMETER, dtls13_cleanup);
+            if (suites.suiteSz > WOLFSSL_MAX_SUITE_SZ)
+                ERROR_OUT(BUFFER_ERROR, dtls13_cleanup);
+            XMEMCPY(suites.suites, ch->cipherSuite.elements, suites.suiteSz);
+
+            /* Populate extensions */
+
+            /* Supported versions always need to be present. Has to appear after
+             * key share as that is the order we reconstruct it in
+             * RestartHandshakeHashWithCookie. */
+            ret = TLSX_Push(&parsedExts,
+                      TLSX_SUPPORTED_VERSIONS, ssl, ssl->heap);
+            if (ret != 0)
+                goto dtls13_cleanup;
```

Comment:
If we encounter an error here, do we reset the hashes? 
It would be nice to compute the cookie without changing `ssl->hsHashes`

Reply:
also, Could you consider refactoring the dtls 1.3  part into a separate function? 
Even if refactored out the function seems quite large. It might be helpful to break it up into smaller functions to make the code easier to read.

Reply:
Refactored to not touch `ssl->hsHashes` while in stateless mode. Refactored 1.3 portion into separate function.